Add assert_precondition helper and PRECONDITION_FAILED subtest status

lukebjerring opened the pull request:
Prototyping/exploring https://github.com/web-platform-tests/rfcs/pull/16

Commit:
Add skip helper

Commit:
Use skip helper

Commit:
Remove file

jimsch commented on resources/testharness.js:
```diff
@@ -1529,7 +1529,8 @@ policies and contribution forms [3].
-        NOTRUN:3
+        NOTRUN:3,
+        SKIP:3
     };
```
Is it intentional that there will not be a distinction between not run and skip?  Might this be desired behavior?

lukebjerring replied:
Not intentional :)

lukebjerring committed:
Update testharness.js

jimsch commented on WebCryptoAPI/generateKey/successes.js:
```diff
@@ -65,7 +65,11 @@ function run_test(algorithmNames, slowTest) {
             }, function(err) {
-                assert_unreached("Threw an unexpected error: " + err.toString());
+                if (err.toString().includes('not supported')) {
+                    test.skip(algorithm.name + ' not implemented');
```
I am wondering if the better thing to do is to use the name of the error rather than assuming that it includes an english string is the correct behavior.  

if (err.name == "NotSupportedError")

lukebjerring replied:
I too wondered this and discovered that `NotSupportedError` is not returned when the algorithm is not supported; it's an `OperationError`, which isn't distinct from the type used for genuine errors.

That's spec compliant, BTW: https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-generateKey

lukebjerring replied:
FWIW, I'm working on getting a less ugly/impractical use-case example from codec tests, in collaboration with @Orphis, so will remove this eye-sore once I have something nicer.

re: NotSupportedError, @ericroman920 informs me that 

> The conclusion was that Web Crypto implementations [must not throw NotImplemented](https://www.w3.org/Bugs/Public/show_bug.cgi?id=27598) (unless they don't support AES at all)

jimsch commented:
I think that I lost the review flag.  If I still had it I would approve.

jgraham commented:
This needs a testharness unittest, and a test in `infrastructure/`. I would expect the latter to `ERROR` until  we get changes in mozlog (and maybe wptrunner) to handle the new status. Since we're using mozlog from PyPI this will also block on getting a new mozlog release.

Hexcles committed:
Add SKIP to a few more enums

`wpt run` works correctly now and can produce JSON with status "SKIP" as
expected.

Hexcles committed:
Add infrastructure test

Hexcles committed:
Add testharness unit test

Hexcles commented:
@jgraham I've fixed a bug and added two tests (infrastructure & resource tests).

jgraham commented:
My only major concern about this is that it makes it impossible to tell the difference between "this test was disabled by the implementation e.g. for stability issues" and "this test was skipped because the feature is unsupported". That difference seems meaningful to me since you might want to audit tests that are disabled for instability,

Commit:
Rename to PRECONDITION_FAILED

lukebjerring commented:
@jgraham - as agreed, switched the (subtest) status name to `PRECONDITION_FAILED`.

(Reasons: Using `SKIP` would be ambiguous with already-supported test-level skippage/disable behaviour, and renaming existing `SKIP` to something else would be a major headache).

Commit:
Change to assert_precondition

Commit:
Fix metadata

lukebjerring commented:
@jgraham - I'm seeing the mozlog error we'd expect:

> https://tools.taskcluster.net/groups/RGFPhLFqRGqvPXAaRXcmGA/tasks/TzNbxq2bTrWslrAGvcfsuA/runs/0/logs/public%2Flogs%2Flive.log#L846

Commit:
Cleanup

lukebjerring committed:
Update tox.ini

foolip commented on WebCryptoAPI/generateKey/successes.js:
```diff
@@ -65,6 +65,8 @@ function run_test(algorithmNames, slowTest) {
                 }
             }, function(err) {
+                var supported = !err.toString().includes('not supported');
+                assert_precondition(supported, algorithm.name + ' not implemented');
```
I'll move the bikeshedding here, since I guess this PR will have the docs and testharness.js changes. What will the resulting message in the report be?

lukebjerring replied:
Ok, can actually answer this now that I've fixed it. It's the message in the second param to `assert_precondition`, e.g. 

![image](https://user-images.githubusercontent.com/777956/66340471-3a1adc00-e913-11e9-9ad3-21d009260594.png)

(bad alignment notwithstanding)

lukebjerring commented on resources/testharness.js:
```diff
@@ -1509,6 +1509,23 @@ policies and contribution forms [3].
+    expose(PreconditionFailedError, "PreconditionFailedError");
+
+    PreconditionFailedError.prototype = Object.create(AssertionError.prototype);
+    PreconditionFailedError.prototype.constructor = PreconditionFailedError;
```
For some reason, `new PreconditionFailedError('foo') instanceof AssertionError` is false.

foolip replied:
The style of inheritence used here looks unfamiliar to me, is it used elsewhere in testharness.js and is `instanceof` correct in those cases?

lukebjerring replied:
It's es5 "inheritance" and yes, `AssertionError` uses it for inheriting `Error`

foolip replied:
It's overriding `Foo.prototype.constructor` that looks unfamiliar. Maybe this pattern just isn't right, or is `new AssertionError('foo') instanceof Error` true?

lukebjerring replied:
So, the issue was that the whole thing is in a closure. It uses `expose` to add the definitions to the global_scope. As a result, the references need to exist before calling expose, otherwise they'll refer to a local of the same name, and the globally resolved `AssertionError` is a different reference, thus instanceof is false. Fixed by declaring the types in the right order.

Commit:
Fix declaration order

lukebjerring commented:
@jgraham @foolip I've run out of time to get this implemented past the testharness.js layer.

Current state of this PR is that a PRECONDITION_FAILED result is successfully produced when loading the test running on `./wpt serve`, but that status is unrecognized (so unrecorded) by the wpt runner / mozlog side.

foolip commented:
@jgraham where is mozlog maintained? I presume this is a pretty simple fix and that most of the work is in the review and release step...

jgraham commented:
It's in mozilla-central: https://searchfox.org/mozilla-central/source/testing/mozbase/mozlog/

foolip commented:
I've filed https://bugzilla.mozilla.org/show_bug.cgi?id=1589056 to track this in mozlog.

foolip commented:
A review is up at https://phabricator.services.mozilla.com/D49450

foolip commented:
I've filed https://github.com/web-platform-tests/wpt/issues/19844 to track this to completion, closing this PR.

foolip commented:
Parts of this was landed in https://github.com/web-platform-tests/wpt/pull/19993 and the test changes are in https://github.com/web-platform-tests/wpt/pull/20173 + https://github.com/web-platform-tests/wpt/pull/20176.
